bittide/src/Bittide/Calculator.hs

```diff
@@ -0,0 +1,571 @@
+-- SPDX-FileCopyrightText: 2025 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Bittide.Calculator where
+
+import Clash.Prelude
+
+import Bittide.SharedTypes (divRU)
+import Clash.Functor.Extra ((<<$>>))
+import Control.Monad (forM_)
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.Map (Map)
+import Data.Maybe (catMaybes, fromJust)
+import Data.String.Interpolate (i)
+import GHC.Stack (HasCallStack)
+
+import qualified Data.List as L
+import qualified Data.Map as Map
+
+type FpgaId = String
+
+-- TODO: find a way to derive this 4 from the code instead of using a magic number.
```

Comment:
```suggestion
-- TODO: find a way to derive this 4 from the code instead of using a magic number.
--
-- This is currently '4', because we traverse two switches and each switch has registered
-- inputs and outputs.
```